src/sctools/encodings.py

```diff
@@ -0,0 +1,167 @@
+    """Encode a DNA sequence using a 2-bit encoding.
+
+    Two-bit encoding uses 0 for an encoded nucleotide. As such, it cannot distinguish between
+    the end of sequence and trailing A nucleotides, and thus decoding these strings requires
+    knowledge of their length. As such, it is only appropriate for encoding fixed sequence
+    lengths
+
+    In addition, in order to encode in 2-bit, N-nucleotides must be randomized to one of A, C,
+    G, and T.
+
+    :param int sequence_length: number of nucleotides that are being encoded
+    """
+
+    def __init__(self, sequence_length):
+        self.sequence_length = sequence_length
+
+    class TwoBitEncodingMap:
+        """
+        custom class that mimics a read-only dict, but generates random nucleotides for ambiguous
+        nucleotides e.g. N
+        """
+        map_ = {65: 0, 67: 1, 84: 2, 71: 3}  # A, C, T, G
+
+        def __getitem__(self, byte):
+            try:
+                return self.map_[byte]
+            except KeyError:
+                if byte != 78:
+                    raise
+                return random.randint(0, 4)
+
+    encoding_map = TwoBitEncodingMap()
+    decoding_map = {0: b'A', 1: b'C', 2: b'T', 3: b'G'}
+    bits_per_base = 2
+
+    @classmethod
+    def encode(cls, bytes_encoded):
+        """
+        encodes a DNA bytes string using 2 bits per base. N (and other ambiguous nucleotides)
+        are randomized
+
+        :param bytes bytes_encoded: a bytes string containing A, C, G, T nucleotides.
+        :return int: encoded sequence
+        """
+        encoded = 0
+        for character in bytes_encoded:
+            encoded <<= 2
+            encoded += cls.encoding_map[character]
+        return encoded
+
+    def decode(self, integer_encoded):
+        """decodes an 2-bit encoded string
+
+        :param int integer_encoded: encoded nucleotide sequence
+        :return bytes: decoded sequence
+        """
+        decoded = b''
+        for _ in range(self.sequence_length):
+            decoded = self.decoding_map[integer_encoded & 3] + decoded
+            integer_encoded >>= 2
+        return decoded
+
+    def gc_content(self, integer_encoded):
+        """return the gc content of an 2-bit encoded bytes object
+
+        :param int integer_encoded: encoded nucleotide string
+        """
+        i = 0
+        for _ in range(self.sequence_length):
+            i += integer_encoded & 1
+            integer_encoded >>= 2
+        return i
+
+
+class ThreeBit(Encoding):
+    """Encode a DNA sequence using a 3-bit encoding.
+
+    Since no bases are encoded as 0, an empty triplet is interpreted as the end of the encoded
+    string; Three-bit encoding can be used to encode and decode strings without knowledge of their
+    length.
+    """
+
+    class ThreeBitEncodingMap:
+
+        # C: 1, A: 2, G: 3, T: 4, N: 6;  # note, not using 0
+        map_ = {65: 2, 97: 2, 67: 1, 99: 1, 71: 3, 103: 3, 84: 4, 116: 4, 78: 6, 110: 6}
```

Comment:
Hmm, unless it would bog down performance, what about creating the map_ this way?
{ord('A'): 2, ord('a'): 2, ... }
That way you are documenting the mapping in code rather than a comment, and less chance for a typo to introduce an error.
Or you could have raw_map = {'A': 2, 'a': 2, ... } and then process that to create map_ = {65: 2, ... }

Reply:
Markus made the same (really good) suggestion. I'm going to change this; much much clearer the way you're both suggesting.